### ISSUE TITLE
LambdarankNDCG objective ignores truncation level

### ISSUE THREAD

**Issue by metpavel** [microsoft/LightGBM#3420]
In /src/objective/rank_objective.hpp: currently, for each query, all pairs of documents contribute to the gradients computation regardless of the truncation level _k_. However, to correctly optimize for NDCG@_k_, one should exclude pairs containing both documents beyond the top _k_ (as they don't affect NDCG@_k_ when swapped). Please enforce the check for the truncation level _k_.

**Comment by guolinke** [comment]
@metpavel during training, some top docs may be wrong ranking, and not in top-k. In these cases, we still need to consider all docs.

**Comment by metpavel** [comment]
@guolinke, in fact, we would still consider all documents even if we exclude pairs containing both documents beyond the top-_k_ (as any document will contribute when paired with any other top-_k_ document -- thus any relevant document has a push towards the top of the list no matter what position it is at). Otherwise, we just optimize for NDCG@Full_Rank regardless of the desired truncation level.

**Comment by guolinke** [comment]
actually, the parameter once named `optimize_pos_at_`,  but I find it actually does not optimize the ndcg@k, and change its name to `truncation_level_`. 

For your solution, `exclude pairs containing both documents beyond the top-k`, I mis-read it. 
I think it is a good solution! 
Do you have the interest to create a PR for this change?
